sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/KeyVaultKeyStore.java

```diff
@@ -0,0 +1,325 @@
+                    alias = candidateAlias;
+                    break;
+                }
+            }
+        }
+        return alias;
+    }
+
+    @Override
+    public Certificate[] engineGetCertificateChain(String alias) {
+        Certificate[] chain = null;
+        Certificate certificate = engineGetCertificate(alias);
+        if (certificate != null) {
+            chain = new Certificate[1];
+            chain[0] = certificate;
+        }
+        return chain;
+    }
+
+    @Override
+    public Date engineGetCreationDate(String alias) {
+        return creationDate;
+    }
+
+    @Override
+    public KeyStore.Entry engineGetEntry(String alias, KeyStore.ProtectionParameter protParam) throws KeyStoreException, NoSuchAlgorithmException, UnrecoverableEntryException {
+        return super.engineGetEntry(alias, protParam);
+    }
+
+    @Override
+    public Key engineGetKey(String alias, char[] password) {
+        Key key;
+        if (certificateKeys.containsKey(alias)) {
+            key = certificateKeys.get(alias);
+        } else {
+            key = keyVault.getKey(alias, password);
+            if (key != null) {
+                certificateKeys.put(alias, key);
+                if (!aliases.contains(alias)) {
+                    aliases.add(alias);
+                }
+            }
+        }
+        return key;
+    }
+
+    @Override
+    public boolean engineIsCertificateEntry(String alias) {
+        if (aliases == null) {
+            aliases = keyVault.getAliases();
+        }
+        return aliases.contains(alias);
+    }
+
+    @Override
+    public boolean engineIsKeyEntry(String alias) {
+        return engineIsCertificateEntry(alias);
+    }
+
+    @Override
+    public void engineLoad(KeyStore.LoadStoreParameter param) {
+        if (param instanceof KeyVaultLoadStoreParameter) {
+            KeyVaultLoadStoreParameter parameter = (KeyVaultLoadStoreParameter) param;
+            keyVault = new KeyVaultClient(
+                parameter.getUri(),
+                parameter.getTenantId(),
+                parameter.getClientId(),
+                parameter.getClientSecret());
+        }
+        sideLoad();
+    }
+
+    @Override
+    public void engineLoad(InputStream stream, char[] password) {
+        sideLoad();
+    }
+
+    @Override
+    public void engineSetCertificateEntry(String alias, Certificate certificate) {
+        if (aliases == null) {
+            aliases = keyVault.getAliases();
+        }
+        if (!aliases.contains(alias)) {
+            aliases.add(alias);
+            certificates.put(alias, certificate);
+        }
+    }
+
+    @Override
+    public void engineSetEntry(String alias, KeyStore.Entry entry, KeyStore.ProtectionParameter protParam) throws KeyStoreException {
+        super.engineSetEntry(alias, entry, protParam);
+    }
+
+    @Override
+    public void engineSetKeyEntry(String alias, Key key, char[] password, Certificate[] chain) {
```

Comment:
Consider throwing UnsupportedOperationException to offer a better User experience, in all the methods that do not have any logic in them.
Currently, user will just wonder if the operation executed successfully or not.

Reply:
The API contract is defined by the JCA specification / JavaDoc and it does not state UnsupportedOperationException can be thrown.